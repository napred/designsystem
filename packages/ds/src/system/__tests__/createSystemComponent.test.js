/**
 * @flow
 * @jest-environment jsdom
 */

import React from 'react';
import styled from 'styled-components';
import { render } from 'enzyme';
import { createSystemComponent, isSystemComponent, DesignSystem } from '../';

describe('createSystemComponent', () => {
  it('creates system component from primitive component', () => {
    const A = createSystemComponent('span');

    expect(isSystemComponent(A)).toBe(true);

    expect(
      render(
        <DesignSystem>
          <A />
        </DesignSystem>,
      ),
    ).toMatchSnapshot();
  });

  it('creates system component from another system component (extends it)', () => {
    const A = createSystemComponent('span');
    const B = createSystemComponent(A);

    expect(isSystemComponent(B)).toBe(true);

    expect(
      render(
        <DesignSystem>
          <B />
        </DesignSystem>,
      ),
    ).toMatchSnapshot();
  });

  it('creates system component from custom component', () => {
    function A(props: any) {
      return <div {...props} />;
    }

    const B = createSystemComponent(A);

    expect(isSystemComponent(B)).toBe(true);

    expect(
      render(
        <DesignSystem>
          <B />
        </DesignSystem>,
      ),
    ).toMatchSnapshot();
  });

  it('does not allow to use styled component as base', () => {
    expect(() => createSystemComponent(styled.div``)).toThrow();
  });

  describe('component overriding', () => {
    it('allows to override base component by primitive component', () => {
      const A = createSystemComponent('span');

      expect(
        render(
          <DesignSystem>
            <A as="div" />
          </DesignSystem>,
        ),
      ).toMatchSnapshot();
    });

    it('allows to override base component by system component (extends it)', () => {
      const A = createSystemComponent('span');
      const B = createSystemComponent('div');

      expect(
        render(
          <DesignSystem>
            <A as={B} />
          </DesignSystem>,
        ),
      ).toMatchSnapshot();
    });

    it('allows to override base component by custom component (not recommended)', () => {
      const A = createSystemComponent('span');
      function B(props: any) {
        return <div {...props} />;
      }

      expect(
        render(
          <DesignSystem>
            <A as={B} />
          </DesignSystem>,
        ),
      ).toMatchSnapshot();
    });

    it('does not allow to use styled component as override', () => {
      const A = createSystemComponent('span');

      expect(() =>
        render(
          <DesignSystem>
            <A as={styled.div``} />
          </DesignSystem>,
        ),
      ).toThrow();
    });
  });
});
