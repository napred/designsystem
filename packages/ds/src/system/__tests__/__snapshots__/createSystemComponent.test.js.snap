// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSystemComponent component overriding allows to override base component by custom component (not recommended) 1`] = `
.c0 {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
}

<div
  blacklist="alignContent,alignItems,alignSelf,bgColor,borderColor,borderRadius,borderRadiusBottomLeft,borderRadiusBottomRight,borderRadiusTopLeft,borderRadiusTopRight,borderStyle,borderWidth,bottom,color,display,flex,flexBasis,flexDirection,flexGrow,flexOrder,flexShrink,flexWrap,fontFamily,fontSize,fontWeight,height,justifyContent,left,lineHeight,m,maxHeight,maxWidth,mb,minHeight,minWidth,ml,mr,mt,mx,my,overflow,overflowX,overflowY,p,pb,pl,position,pr,pt,px,py,right,shadow,system,textAlign,textOverflow,textTransform,top,transition,viewport,whiteSpace,width,zIndex"
  class="c0"
  system="[object Object]"
  viewport="0"
/>
`;

exports[`createSystemComponent component overriding allows to override base component by primitive component 1`] = `
.c0 {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
}

<div
  class="c0"
/>
`;

exports[`createSystemComponent component overriding allows to override base component by system component (extends it) 1`] = `
.c0 {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
}

<div
  class="c0"
/>
`;

exports[`createSystemComponent creates system component from another system component (extends it) 1`] = `
.c0 {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
}

<span
  class="c0"
/>
`;

exports[`createSystemComponent creates system component from custom component 1`] = `
.c0 {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
}

<div
  blacklist="alignContent,alignItems,alignSelf,bgColor,borderColor,borderRadius,borderRadiusBottomLeft,borderRadiusBottomRight,borderRadiusTopLeft,borderRadiusTopRight,borderStyle,borderWidth,bottom,color,display,flex,flexBasis,flexDirection,flexGrow,flexOrder,flexShrink,flexWrap,fontFamily,fontSize,fontWeight,height,justifyContent,left,lineHeight,m,maxHeight,maxWidth,mb,minHeight,minWidth,ml,mr,mt,mx,my,overflow,overflowX,overflowY,p,pb,pl,position,pr,pt,px,py,right,shadow,system,textAlign,textOverflow,textTransform,top,transition,viewport,whiteSpace,width,zIndex"
  class="c0"
  system="[object Object]"
  viewport="0"
/>
`;

exports[`createSystemComponent creates system component from primitive component 1`] = `
.c0 {
  font-family: "Palanquin","Helvetica",sans-serif;
  font-size: 16px;
}

<span
  class="c0"
/>
`;
